contextual.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package klog
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+)
+
+// This file provides the implementation of
+// https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging
+//
+// SetLogger and ClearLogger were originally added to klog.go and got moved
+// here. Contextual logging adds a way to retrieve a Logger for direct logging
+// without the logging calls in klog.go.
+//
+// The global variables are expected to be modified only during sequential
+// parts of a program (init, serial tests) and therefore are not protected by
+// mutex locking.
+
+var (
+	// contextualLoggingEnabled controls whether contextual logging is
+	// active. Disabling it may have some small performance benefit.
+	contextualLoggingEnabled = true
+
+	// globalLogger is the global Logger chosen by users of klog, nil if
+	// none is available.
+	globalLogger *Logger
+
+	// contextualLogger defines whether globalLogger may get called
+	// directly.
+	contextualLogger bool
+
+	// klogLogger is used as fallback for logging through the normal klog code
+	// when no Logger is set.
+	klogLogger logr.Logger = logr.New(&klogger{})
+)
+
+// SetLogger sets a Logger implementation that will be used as backing
+// implementation of the traditional klog log calls. klog will do its own
+// verbosity checks before calling logger.V().Info. logger.Error is always
+// called, regardless of the klog verbosity settings.
+//
+// If set, all log lines will be suppressed from the regular Output, and
+// redirected to the logr implementation.
+// Use as:
+//   ...
+//   klog.SetLogger(zapr.NewLogger(zapLog))
+//
+// To remove a backing logr implemention, use ClearLogger. Setting an
+// empty logger with SetLogger(logr.Logger{}) does not work.
+//
+// Modifying the logger is not thread-safe and should be done while no other
+// goroutines invoke log calls, usually during program initialization.
+func SetLogger(logger logr.Logger) {
+	globalLogger = &logger
+	contextualLogger = false
+}
+
+// SetContextualLogger does the same as SetLogger, but in addition the
+// logger may also get called directly by code that retrieves it
+// with FromContext, TODO or Background. The logger therefore must
+// implements its own verbosity checking.
+func SetContextualLogger(logger logr.Logger) {
```

Comment:
Why have two different way to switch `contextualLogger` on? I think that having one switch like `EnableContextualLogging` is enough. Having two ways makes creates inconsistency and needless discussion on which option should be used.

Reply:
It's just a feeling, but replacing `klog.SetContextualLogger(logger)` with
```
klog.SetLogger(logger)
klog.EnableContextualLogging(true)
```
doesn't send a clear message to me that `logger` must meet certain criteria for this to work.

Also, `EnableContextualLogging` as it is defined right now is about something else and cannot be used to enable the additional usage of the logger.

It's on by default and controls the behavior of calls like `FromContext`. It's meaning is not "the logger can be used as contextual logger". We would need a separate `klog.UseLoggerAsContextualLogger` call for that - or something like that. I can't think of a good name and prefer to keep `SetContextualLogger`.

Reply:
> It's just a feeling, but replacing `klog.SetContextualLogger(logger)` with
> 
> ```
> klog.SetLogger(logger)
> klog.EnableContextualLogging(true)
> ```
> 
> doesn't send a clear message to me that `logger` must meet certain criteria for this to work.

If I remember this criteria, it was that provided `logger` needs to execute verbosity checks for itself. Right? I think the `logr` interface already expects that. I don't think we should be guided by rushed implementation for K8s json logging that skipped this. 

Just want to make sure we don't bring old baggage to the API.

Reply:
> If I remember this criteria, it was that provided logger needs to execute verbosity checks for itself. Right?

Correct.

> I don't think we should be guided by rushed implementation for K8s json logging that skipped this.

We can't be sure that other implementations didn't do the same. Searching for it with https://grep.app/search?q=klog.SetLogger finds several places that'll get broken if we change the requirements for the logger passed to SetLogger:

* https://github.com/projectcontour/contour/blob/d231033fc25e8caf1137282ae8d49cbab562160f/internal/k8s/log.go#L78-L83
* https://github.com/cri-o/cri-o/blob/f1d3a46642377d9bf4cb79255020b07578023540/internal/log/klog.go#L77

Reply:
> I think the logr interface already expects that.

The interface only expects that there is an Enabled function. It does not specify a specific semantic like "compare against your verbosity threshold", therefore an implementation that always returns true is valid. klog did specify a semantic ("verbosity check done by klog, logger should always log what it gets") and we shouldn't change that for `SetLogger`.

Reply:
I think we should focus on what is a good long term API and treat the breaking change as a separate problem. I don't think having two methods for setting logger is good and we should have a plan to unify them at some point. If we are worried about breaking klog users, then we should make sure that they handle the verbosity. As this will take some time I propose to:
* Make `SetContextualLogger` an experimental/temporary function that will be removed.
* Identify and file issue to github projects that didn't implement proper verbosity checking.
* Remove `SetContextualLogger` when all the cases are fixed.

Still it maybe to much work.

Reply:
I think it will create work for us and the ecosystem that isn't justified.

@dims: what do you prefer?

Reply:
@pohly @serathius Alternative - let's start from `func SetLogger(logger logr.Logger)` what if the `logger` parameter can implement an addition interface/method say `SupportsContext` which returns bool (if true, it will be the same as `SetContextualLogger` as in the PR currently. if false or if this interface/method is absent, it will mean the current behavior of `SetLogger`)

If we do this,
- Current code does not have to change for anyone (the 2 instances pointed out in cri-o and contour)
- Folks can explicitly opt in when they want to do the new thing
- We avoid busy work of adding/removing a temporary method etc.

Does this solve the problem or am i way off base? thanks!

Reply:
That means one cannot use some existing Logger implementation unchanged with klog. One can wrap an existing one to add that function and that wrapper itself can exist in klog, but is that really simpler than a second function? Code:

```
func SetLogger(logger logr.Logger) {
    // interface check here
}

func AddSupportsContext(logger logr.Logger) logr.Logger {
     return wrapper{logger}
}

type wrapper struct {
   logr.Logger
}

func (w wrapper) SupportsContext() bool {
    return true
}

// usage:

klog.SetLogger(klog.AddSupportsContext(logger))
```

Reply:
Yes, i prefer the explicit opt-in rather than a second method.

Reply:
I started implementing this and then found that this cannot work. `logr.Logger` is not an interface anymore since logr v1.0, therefore it is impossible to add additional methods. We would have to modify logr, but such a `SupportsContext` function doesn't belong into that API.

Reply:
hmm. i keep forgetting that. how about the `LogSink`? roughly

```
func SetLogger(logger logr.Logger) {
    // call logger.GetSink() and do the
    // interface/method check here
}

func AddSupportsContext(logger logr.Logger) logr.Logger {
     sink := logger.GetSink()
     return logger.WithSink(wrapper{logger})
}

type wrapper struct {
   logr.LogSink
}

func (w wrapper) SupportsContext() bool {
    return true
}

// usage:

klog.SetLogger(klog.AddSupportsContext(logger))
```

Reply:
That works. See additional commit for the result.

I find it rather ugly and it's also fragile (wrapping doesn't survive a logger.WithValues, for example), but it gets the job done. I can live with this.

Reply:
As I noticed while working on this, wrapping like this is dropping the hidden `logr.Logger.level` field. To me this looks more and more like a hack.

Reply:
> * Make SetContextualLogger an experimental/temporary function that will be removed.
> * Identify and file issue to github projects that didn't implement proper verbosity checking.
> * Remove SetContextualLogger when all the cases are fixed.

That means whoever wants two use contextual logging and directly call the logger must migrate *twice*: first to SetContextualLogger because SetLogger will not do what they want, then back to SetLogger once we have changed it. That doesn't sound appealling.

Another alternative is:
- introduce SetContextualLogger as experimental (state of this PR before removing it)
- when we consider it stable, remove the "experimental" tag
- at the same time or later, mark `SetLogger` as deprecated
- remove it when it no longer seems to be used

I don't know whether we can do that within the klog/v2. I'm also not sure whether "clean API" is worth all the trouble for the community.

Yet another alternative: adding optional functional parameters to SetLogger, without adding SetContextualLogger. The only parameter will be `ContextualLogger(enabled bool)`. This is an API break in theory (function signature changes, so code which uses it for a function pointer breaks), but in practice the only usage seem to be direct calls and those continue to work without changes.

Reply:
> Another alternative is: [...] remove SetLogger when it no longer seems to be used

This implies that downstream users must change their code although it currently works fine for them. If they are happy with the current approach (klog flags + verbosity checks, Logger without verbosity check as backend), why should they change? Do we gain that much from the removal that it is worth breaking working code? I don't think so.